torch/csrc/Device.cpp

```diff
@@ -1,235 +1,106 @@
-  PyVarObject_HEAD_INIT(nullptr, 0)
-  "torch.device",                        /* tp_name */
-  sizeof(THPDevice),                     /* tp_basicsize */
-  0,                                     /* tp_itemsize */
-  nullptr,                               /* tp_dealloc */
-  0,                                     /* tp_vectorcall_offset */
-  nullptr,                               /* tp_getattr */
-  nullptr,                               /* tp_setattr */
-  nullptr,                               /* tp_reserved */
-  (reprfunc)THPDevice_repr,              /* tp_repr */
-  nullptr,                               /* tp_as_number */
-  nullptr,                               /* tp_as_sequence */
-  nullptr,                               /* tp_as_mapping */
-  (hashfunc)THPDevice_hash,              /* tp_hash  */
-  nullptr,                               /* tp_call */
-  (reprfunc)THPDevice_str,               /* tp_str */
-  nullptr,                               /* tp_getattro */
-  nullptr,                               /* tp_setattro */
-  nullptr,                               /* tp_as_buffer */
-  Py_TPFLAGS_DEFAULT,                    /* tp_flags */
-  nullptr,                               /* tp_doc */
-  nullptr,                               /* tp_traverse */
-  nullptr,                               /* tp_clear */
-  (richcmpfunc)THPDevice_rc,             /* tp_richcompare */
-  0,                                     /* tp_weaklistoffset */
-  nullptr,                               /* tp_iter */
-  nullptr,                               /* tp_iternext */
-  THPDevice_methods,                     /* tp_methods */
-  nullptr,                               /* tp_members */
-  THPDevice_properties,                  /* tp_getset */
-  nullptr,                               /* tp_base */
-  nullptr,                               /* tp_dict */
-  nullptr,                               /* tp_descr_get */
-  nullptr,                               /* tp_descr_set */
-  0,                                     /* tp_dictoffset */
-  nullptr,                               /* tp_init */
-  nullptr,                               /* tp_alloc */
-  THPDevice_pynew,                       /* tp_new */
-};
-
-void THPDevice_init(PyObject *module)
-{
-  if (PyType_Ready(&THPDeviceType) < 0) {
-    throw python_error();
-  }
-  Py_INCREF(&THPDeviceType);
-  if (PyModule_AddObject(module, "device", (PyObject *)&THPDeviceType) != 0) {
-    throw python_error();
-  }
+void initDeviceBindings(PyObject* module) {
+  py::options options;
+  options.disable_user_defined_docstrings();
+  options.disable_function_signatures();
+
+  // NB: If you edit these properties/methods, update torch/_C/__init__.pyi.in
+  py::class_<Device>(module, "device", py::is_final())
+      .def(py::init<const Device&>(), py::arg("device"))
+      .def(py::init(wrap_pybind_function(parseDevice)), py::arg("device"))
+      .def(py::init(
+          [](const std::string& type, DeviceIndex index) {
```

Comment:
nit: for consistency, can you lump this case in with the others in `parseDevice`? It looks like that function deals with all of the other constructor variants

Reply:
Well that one is intentional... The copy constructor for `at::Device` never fails, so we create a separate init implementation for it to avoid `HANDLE_TH_ERRORS` and `END_HANDLE_TH_ERRORS_PYBIND`.

Reply:
ah yep, missed that.